eZ/Publish/Core/Repository/ComparingService.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\Repository;
+
+use eZ\Publish\API\Repository\ComparingService as ComparingServiceInterface;
+use eZ\Publish\API\Repository\Values\Content\VersionDiff\FieldDiff;
+use eZ\Publish\API\Repository\Values\Content\VersionDiff\VersionDiff;
+use eZ\Publish\API\Repository\Values\Content\VersionInfo;
+use eZ\Publish\Core\Compare\CompareEngineRegistry;
+use eZ\Publish\Core\Compare\FieldRegistry;
+use eZ\Publish\Core\Repository\Helper\ContentTypeDomainMapper;
+use eZ\Publish\SPI\Persistence\Content;
+use eZ\Publish\SPI\Persistence\Content\Field;
+use eZ\Publish\SPI\Persistence\Content\Handler as ContentHandler;
+use eZ\Publish\SPI\Persistence\Content\Type;
+use eZ\Publish\SPI\Persistence\Content\Type\Handler as ContentTypeHandler;
+
+class ComparingService implements ComparingServiceInterface
+{
+    /** @var \eZ\Publish\SPI\Persistence\Content\Handler */
+    private $contentHandler;
+
+    /** @var \eZ\Publish\SPI\Persistence\Content\Type\Handler */
+    private $contentTypeHandler;
+
+    /** @var \eZ\Publish\Core\Compare\FieldRegistry */
+    private $fieldRegistry;
+
+    /** @var \eZ\Publish\Core\Compare\CompareEngineRegistry */
+    private $comparatorEngineRegistry;
+
+    /** @var \eZ\Publish\Core\Repository\Helper\ContentTypeDomainMapper */
+    private $contentTypeDomainMapper;
+
+    public function __construct(
+        ContentHandler $contentHandler,
+        ContentTypeHandler $contentTypeHandler,
+        FieldRegistry $fieldRegistry,
+        CompareEngineRegistry $comparatorEngineRegistry,
+        ContentTypeDomainMapper $contentTypeDomainMapper
+    ) {
+        $this->contentHandler = $contentHandler;
+        $this->contentTypeHandler = $contentTypeHandler;
+        $this->fieldRegistry = $fieldRegistry;
+        $this->comparatorEngineRegistry = $comparatorEngineRegistry;
+        $this->contentTypeDomainMapper = $contentTypeDomainMapper;
+    }
+
+    public function compareVersions(
+        VersionInfo $versionA,
+        VersionInfo $versionB,
+        ?string $languageCode = null
+    ): VersionDiff {
+        $content = $this->contentHandler->load($versionA->getContentInfo()->id, $versionA->versionNo);
+        $contentToCompare = $this->contentHandler->load($versionB->getContentInfo()->id, $versionB->versionNo);
```

Comment:
Should we somewhere check if both versions belong to the same content? Or it should be possible to compare versions for different contents?

Reply:
Good question. If allowing to compare versions of different contents, should we check 1st that they share the same Type? Or shall we allow cross-type checks, which is nice but adds some complexity to the comparison logic (eg. look for fields in A but not in B, look in fields in B but not in A, look for fields in both A and B but having different type, compare the rest ?)

Reply:
I have added check for same content. I think comparing across contents can be done on field level? Like define which fields you want to compare, but I would add this later if need as another method as I think this is complicated enough already.

Reply:
why not start with a check on same Type instead of same Content ? Seems like same complexity, broader usecase

Reply:
@gggeek I don't see a use case for comparing the 2 different content items (same type), do you?

Reply:
In general? I think it is a quite useful feature

Reply:
@gggeek I am not `compare version` is the best interface to identify duplicated images.
We have a feature in our backlog that can be used for that https://portal.productboard.com/ezproduct/1-ez-platform-roadmap/c/123-identify-duplicate-image-assets-when-creating-an-image-asset-field

Reply:
@SylvainGuittard you might be right.

However, I think that (short version):
- the change to allow comparing 2 versions on different contents of same type in terms of code is probably a couple of lines. Not a lot of work (of course you might want to write one test scenario for this case so let's say a _little bit_ of work)
- there is no downside that I can think of about adding that feature to the API
- almost certainly, end users will find useful cases for that feature
- the feature 'find duplicate image assets' is not generic enough, at least reading from its title; it could be applied to any binary-file content as well, eg. "find duplicate pdf files"

As for the long version:

some of the defining characteristics of eZPublish4 are that it was a _flexible_ and _open_ system, which used heavily _common patterns_ across its API layers. This means that developers were free to extend and expand the system with ease, once they had grasped the few basic concepts. Overrides worked similarly for settings and design elements; templates could be overridden the same way for Content display as well as for Field display; everything-is-a-content; backoffice customization was the same as frontend customization.
Some of the less-successful (or missing) features of the system came in fact from the fact that they were implemented ad-hoc for some specific usages instead of being made available as generic Content Repository functions. Eg. eZFlow way of scheduling content publication never could cope with scheduled publication of non-ezflow contents.

The above is to say that, although it is impossible to have 100% clear vision on future needs, and that there is a cost to implementing APIs that are too far-and-wide and flexible, esp. when there is no clearly defined interface or consumer to use them yet, there is also a huge cost in implementing APIs that are short-sighted and based one-to-one to end-user-visible features: that will end up in:
- more APIs, which means
- more code, ie. more maintenance
- more cognitive burden for the developers
- more churn, as changes in the requirements of user-visible features will have to be reflected in the underlying APIs
- which means more breackage for users of the APIs

In my (not so) humble opinion, a good API design is
- layered: the closer the API is to the core, the more it has to be generic. Specific apis sit on top of a pyramid, and make use of the generic ones
- informed: whenever a feature is sorted out by Product Management for implementation, a brain storming session should be had with power users, senior integrators, consultants. Because _they_ are collectively the ones who know what they needed in past projects, and they are the ones that can suggest a small change to an api that will result in it having a much more meaningful impact
- uniform: to reduce cognitive burden and the learning curve, we should use similar names for similar things (and viceversa), and reuse as much as possible the same pattern everywhere
- visionary: well, hard to define and to achieve. But we could probably agree on what a visionary api designer looks like

In concrete terms, an example of an API proposal that seems to me to fall into the "not flexible enough" case: '[RFC] Location references' (# 2766)

Thx for reading all of this :-)
Gaetano